src/security.ts

```diff
@@ -0,0 +1,53 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import traverse from 'traverse'
+
+const secretKeys = [
+  'authorization',
+  'password',
+  'apikey',
+  'x-elastic-app-auth'
+]
+
+/**
+ * Clones an object and recursively loops through all keys, redacting their values if the key matches any of a list of strings.
+ * @param obj: Object to clone and redact
+ * @param additionalKeys: Extra keys that can be matched for redaction. Does not overwrite the default set.
+ */
+export function redactObject (obj: Record<string, any>, additionalKeys: string[] = []): Record<string, any> {
+  const toRedact = [...secretKeys, ...additionalKeys].map(key => key.toLowerCase())
+  return traverse(obj)
```

Comment:
[`traverse`](https://www.npmjs.com/package/traverse) loops recursively through every key/value in any deeply-nested object, and can easily account for circular references (see line 47). I wanted both of these things so that this function could be generic enough to be used elsewhere later.

It was proposed that we could use [`fast-redact`](https://www.npmjs.com/package/fast-redact) to implement this redaction feature. However, it uses [wildcards](https://www.npmjs.com/package/fast-redact#wildcards) to match a key string at multiple levels. I benchmarked it against my solution and, when using it to match keys at 5+ levels of nesting&mdash;`DiagnosticResult` includes HTTP headers nested 5 levels deep&mdash;it was substantially slower than using `traverse`, and did not account for circular references.